### PR TITLE
Add dirty rectangle tracking to Shape display element

### PR DESCRIPTION
**Pull request by kmatch98**
Resolves issue: https://github.com/adafruit/circuitpython/issues/3201

### CODE REVIEW

File: shared-module/displayio/Shape.c [hunk 1]
```diff
@@ -49,10 +49,16 @@ void common_hal_displayio_shape_construct(displayio_shape_t *self, uint32_t widt
     self->half_height = height;
 
     self->data = m_malloc(height * sizeof(uint32_t), false);
+    //for (uint16_t i = 0; i < height; i++) {
```
Comment (FoamyGuy) [inline shared-module/displayio/Shape.c:52]: Do we need to keep this commented out line?

**Review by FoamyGuy — COMMENTED** [review]
I tested this on a CLUE and PyPortal using the code from the issue. This version does fix the problem on both devices with that code as it's written. 

I noticed on the CLUE I could get back to non-refreshing state still by using lower values for the `time.sleep()`. I do think there is still something else going on there somehere.

The PyPortal seemed to handle it fine even at faster speeds that didn't work on the CLUE. 

This is great extra functionality to be able to resize the Shape objects. Thanks for working on it @kmatch98 !

File: shared-module/displayio/Shape.c (hunk 1 above)
Comment (kmatch98) [reply shared-module/displayio/Shape.c]: Deleted unnecessary comment.  Thanks for testing it and catching this leftover comment.

File: shared-module/displayio/Shape.c [hunk 2]
```diff
@@ -49,10 +49,16 @@ void common_hal_displayio_shape_construct(displayio_shape_t *self, uint32_t widt
...
 
     self->data = m_malloc(height * sizeof(uint32_t), false);
+
     for (uint16_t i = 0; i <= height; i++) {
         self->data[2 * i] = 0;
         self->data[2 * i + 1] = width;
     }
+
+    self->dirty_area.x1=0;
+    self->dirty_area.x2=width;
```
Comment (tannewt) [inline shared-module/displayio/Shape.c:59]: I don't think you want any changes by default.

```suggestion
    self->dirty_area.x2=0;
```

File: shared-module/displayio/Shape.c [hunk 3]
```diff
@@ -66,8 +72,58 @@ void common_hal_displayio_shape_set_boundary(displayio_shape_t *self, uint16_t y
...
     }
+
+    uint16_t lower_x, upper_x;
+
+    // find x-boundaries for updating based on current data and start_x, end_x
+    if (start_x < self->data[2 * y]) {
+        lower_x = start_x;
+    } else {
+        lower_x = self->data[2 * y];
+    }
```
Comment (tannewt) [inline shared-module/displayio/Shape.c:83]: Mind rewriting this bit with MIN and MAX macros from `py/misc.h`? I think it'll make the code much clearer.

**Review by tannewt — CHANGES_REQUESTED** [review]
Good job handling the mirroring and hooking into the refresh API. The core logic is a bit hard to read so I've suggested using helper macros to make it clearer. Thanks!

File: shared-module/displayio/Shape.c (hunk 3 above)
Comment (kmatch98) [reply shared-module/displayio/Shape.c]: The code was even crazier when I didn't use these intermediate `upper_x` and `lower_x` variables!

Thanks for pointing out the availability of MAX/MIN in `py.misc.h`.  I incorporated these changes.

File: shared-module/displayio/Shape.c (hunk 2 above)
Comment (kmatch98) [reply shared-module/displayio/Shape.c:60]: When the Shape is constructed, I think the dirty area should cover the whole Shape.  
I wrote it consistent with the constructor for `Bitmap.c` (see below).  Maybe I am missing something?

https://github.com/adafruit/circuitpython/blob/bbd802940ee707fbf8647e0f1017dfb713a94c14/shared-module/displayio/Bitmap.c#L67-L70

**Also, I feel like the `Shape` code has an "off by one" error in the for loop in line 53.**  The malloc is based on `height`, but the for loop here goes from `0` to `height` inclusive (which is `height+1` times).  I welcome your feedback.

Comment (tannewt) [reply shared-module/displayio/Shape.c:60]: Ah ok. Yup, I think you are right about the off by one error.

File: shared-module/displayio/Shape.c [hunk 4]
```diff
@@ -66,8 +73,48 @@ void common_hal_displayio_shape_set_boundary(displayio_shape_t *self, uint16_t y
...
+
+    if (self->dirty_area.x1 == self->dirty_area.x2) { // Dirty region is empty
+        self->dirty_area.x1=lower_x;
+        self->dirty_area.x2=upper_x;
+        self->dirty_area.y1 = y;
+        if (self->mirror_y) {
+            self->dirty_area.y2 = self->height-y;
+        } else {
+            self->dirty_area.y2 = y+1;
+        }
```
Comment (tannewt) [inline shared-module/displayio/Shape.c:99]: I think this is correct but I'd be simpler if you handled computing y bounds before saving it into dirty_area.

So, factor it out so you have lower_y and upper_y and then do the merge with the existing area.

**Review by tannewt — CHANGES_REQUESTED** [review]
Better and better! Just a couple more things. Thanks!

File: shared-module/displayio/Shape.c (hunk 4 above)
Comment (kmatch98) [reply shared-module/displayio/Shape.c]: Ok, thanks for the additional prodding, I now see what you were getting at.  I cleaned up the y-portions similar to what I did with x, I hope it's clearer.

Also, I fixed the off-by-one error and realized that the dirty-rectangle tracking is "exclusive" for the `.x2` and `.y2` values, so I plastered that through the comments for posterity.  I think this is good now.

File: shared-module/displayio/Shape.c (hunk 2 above)
Comment (kmatch98) [reply shared-module/displayio/Shape.c:60]: I fixed the off-by-one error and corrected an issue with `half_width` and `half_height` calculation.

Also, I clarified that the dirty rectangle tracking is exclusive for `.x2` and `.y2`.

**Review by tannewt — APPROVED** [review]
Much clearer! Thank you!

**Pull request merged**